Handle PR/MR updates in the SCM integration

saraycp opened the pull request:
Every time a Pull Request is updated, for example, the user adds a new commit to the existing PR, GitHub is going to send the event `pull_request` with `synchronize` action. We start handling those events.

As the package is probably already branched, we just need to ensure that:
- the source code gets updated and the build is retriggered, as soon as the `_branch_request` file gets updated;
- ~we cancel the jobs related to the previous update (or the creation of the PR);~
- we updated the payload on the corresponding `EventSubscriptions` since as soon as the build finish, they have to report to the most recent `commit SHA`.

hennevogel commented on src/api/app/models/workflow/step/branch_package_step.rb:
```diff
@@ -11,18 +11,24 @@ def initialize(step_instructions:, scm_extractor_payload:)
-        branched_package = branch
-        return unless branched_package
+        if updated_pull_request? && target_package.exists?
+          # Having _service in the branched package, it will be updated and will rebuild automatically
```
Hm, who will update it?

dmarcoux commented:
I think some of the commits from this PR could be split out in smaller PRs. It would be easier to review and merge the changes.

vpereira commented:
> I think some of the commits from this PR could be split out in smaller PRs. It would be easier to review and merge the changes.

there are 2 commits. Do you want to break them, each one in one PR?

saraycp commented:
> > I think some of the commits from this PR could be split out in smaller PRs. It would be easier to review and merge the changes.
> 
> there are 2 commits. Do you want to break them, each one in one PR?

@vpereira I squashed the commits, rebased on master, and resolved the conflicts.
I also adapted some code to avoid Code Climate complaints and replaced `User.session` with `@token.user` to make up the `target_project_name`.

danidoni commented on src/api/app/services/trigger_controller_service/scm_extractor.rb:
```diff
@@ -11,6 +11,8 @@ def initialize(scm, event, payload)
     end
 
+    # TODO: this check seems redundant, but it prevents extracting the payload if the event and action
+    # are not the admited ones. See BranchPackageStep.
```
```suggestion
    # are not the allowed ones. See BranchPackageStep.
```

danidoni replied:
:thought_balloon:

dmarcoux commented:
Regarding the conflict in `src/api/app/models/workflow/step/branch_package_step.rb`, this is due to the changes from #11225. 

The methods `source_project`, `source_package`, `target_project` and `target_package` are now public since we depend on them to be able to report back to the SCM.

dmarcoux commented on src/api/app/models/token/workflow.rb:
```diff
@@ -29,10 +29,10 @@ def run_step_and_report(step, scm_extractor_payload, scm_token)
       # TODO: Fix n+1 queries
       repository.architectures.each do |architecture|
-        SCMStatusReporter.new({ project: step.target_project, package: step.target_package, repository: repository.name, arch: architecture.name },
+        SCMStatusReporter.new({ project: step.target_project_name, package: step.target_package, repository: repository.name, arch: architecture.name },
```
One renaming is missing:

```suggestion
        SCMStatusReporter.new({ project: step.target_project_name, package: step.target_package_name, repository: repository.name, arch: architecture.name },
```

saraycp replied:
done!

dmarcoux reviewed (changes requested):
There is also a typo in the second commit message:

Adjust Bra**n**chPackageStep model specs

hennevogel commented on src/api/app/models/workflow/step/branch_package_step.rb:
```diff
@@ -14,43 +14,55 @@ def initialize(step_instructions:, scm_extractor_payload:, token:)
+      end
+
+      def create_or_update_branched_package
+        return target_package if updated_pull_request? && target_package.present?
```
No one is touching the branched package. That means it won't rebuild right?

hennevogel replied:
Also the `branch_request_file` contains the `sha` that won't be updated...

saraycp replied:
I've adapted the BranchPackageStep code. When the PR/MR is updated, the `_branch_request` file is updated too, so it contains the new commit SHA. This way the source code gets updated and the build is triggered again.

I also adapted the specs accordingly.

danidoni commented on src/api/app/models/workflow/step/branch_package_step.rb:
```diff
@@ -95,7 +105,12 @@ def new_pull_request?
+      def add_or_update_branch_request_file(package:)
         branch_request_file = case @scm_extractor_payload[:scm]
                               when 'github'
                                 branch_request_file_github
```
What about a better name like `branch_request_file_options_for_github`?

saraycp replied:
I've used `branch_request_content_for_github` instead, since what the method returns is going to be the content of the file.

danidoni commented on src/api/app/models/workflow/step/branch_package_step.rb:
```diff
@@ -14,43 +14,53 @@ def initialize(step_instructions:, scm_extractor_payload:, token:)
+      def find_or_create_branched_package
+        return target_package if updated_pull_request? && target_package.present?
+
+        branched_package = branch
+        return branched_package if branched_package
```
```suggestion
        branch # returns a branched package or nil, so the if is not really needed
```

danidoni replied:
returns a branched package or nil, so the if is not really needed

saraycp replied:
done
